pylearn2/linear/cudnn2d.py

```diff
@@ -0,0 +1,195 @@
+import functools
+import numpy as np
+
+from theano.sandbox.cuda.dnn import GpuDnnConv, GpuDnnConvDesc
+from theano.sandbox.cuda.basic_ops import gpu_contiguous
+
+from pylearn2.packaged_dependencies.theano_linear.conv2d \
+    import Conv2d as OrigConv2D
+
+from pylearn2.linear.linear_transform import LinearTransform as P2LT
+from pylearn2.utils import sharedX
+from pylearn2.utils.rng import make_np_rng
+
+
+default_seed = [2012, 11, 6, 9]
+default_sparse_seed = [2012, 11, 6]
+
+
+class Cudnn2D(OrigConv2D):
+    """
+    Wrapper on the Theano Cudnn op.
+
+    Parameters
+    ----------
+    filters : Theano shared variable
+        4D-tensor of shape (out channels, in channels, rows, cols)
+    batch_size : int
+        The size of the input batches
+    input_space : Space
+        The Space of the input data
+    output_axes : tuple, optional
+        The requested output axes. If not specified `bc01` will be used.
+    subsample : tuple or list, optional
+        Factor by which to subsample the output. Default (1, 1)
+    border_mode : string, optional
+        ‘valid’ or ‘full’. See scipy.signal.convolve2d
+    filters_shape : tuple of length 2 or 3, optional
+        ([filter's number,] filter's height, filter's width)
+    message : string, optional
+        TODO
+    """
+
+    def __init__(self,
+                 filters,
+                 batch_size,
+                 input_space,
+                 output_axes=('b', 'c', 0, 1),
+                 subsample=(1, 1),
+                 border_mode='valid',
+                 filters_shape=None,
+                 message=''):
+
+        assert batch_size is None or batch_size > 0
+        self._input_space = input_space
+        self._output_axes = output_axes
+        self._subsample = tuple(subsample)
+        self._filters = filters
+
+        super(Cudnn2D, self).__init__(
+            filters=filters,
+            img_shape=(batch_size, input_space.num_channels,
+                       input_space.shape[0], input_space.shape[1]),
+            subsample=self.subsample,
+            border_mode=border_mode,
+            filters_shape=filters.get_value(borrow=True).shape,
+            message=message
+        )
+
+        # conv_op has to be changed
+        self._conv_op = GpuDnnConv()
+        self._desc = GpuDnnConvDesc(border_mode=border_mode,
+                                    subsample=self.subsample,
+                                    conv_mode='conv')
+
+    @functools.wraps(P2LT.get_params)
+    def get_params(self):
+        return [self._filters]
+
+    @functools.wraps(P2LT.get_weights_topo)
+    def get_weights_topo(self, borrow):
+        """
+        Parameters
+        ----------
+        borrow : TODO
+            TODO
+        """
+        return np.transpose(self._filters.get_value(borrow=borrow),
+                            (0, 2, 3, 1))
```

Comment:
This seems to assume filters is provided with a particular axes order. Can you confirm?
In this case I guess it would be better to generalize to whatever input axes where provided.